Updated brer_plugin to support premature advancement fixes in run_brer

carolinedav opened the pull request:
Placed suggestions into `brer_plugin`. Question: Do all suggestions in #19 give a full solution? `linearstoppotential.cpp` has additional lines: 
```c++
params->alpha = alpha;
  params->tolerance = tolerance;
  params->target = target;
  params->samplePeriod = samplePeriod;
  params->logging_filename = logging_filename;
  ```
  
  and we should add `params->stop_called = stop+called;` ?

carolinedav committed:
Updated brer_plugin to support premature advancement fixes in run_brer

eirrgang commented:
> Placed suggestions into `brer_plugin`. Question: Do all suggestions in #19 give a full solution? `linearstoppotential.cpp` has additional lines:
> 
> ```c++
> params->alpha = alpha;
>   params->tolerance = tolerance;
>   params->target = target;
>   params->samplePeriod = samplePeriod;
>   params->logging_filename = logging_filename;
> ```
> 
> and we should add `params->stop_called = stop+called;` ?

I think you're referring to the initialization (but please link to a specific line if I'm wrong). I think we need to initialize with `stop_called = false;`. It makes sense, and it is necessary for correct ensemble behavior (if/where gmxapi is managing multiple coscheduled simulations).

carolinedav commented:
I changed `stop_not_called=true` to `stop_called=false`, but was wondering if we needed to include `stop_called` in the greater [class](https://github.com/kassonlab/brer_plugin/blob/baa32f0909314c21461ad30f973c82c0f3118e26/src/cpp/linearstoppotential.h#L44) by including `stop_called` along with `alpha`, `target`, `tolerance`, etc.

eirrgang commented on src/cpp/linearstoppotential.h:
```diff
@@ -77,7 +78,7 @@ class LinearStop {
   std::string logging_filename_;
   std::unique_ptr<RAIIFile> logging_file_{nullptr};
-  bool stop_not_called_{true};
+  bool stop_called{false};
```
Let's keep consistent with the value referenced in line 61 and with the GROMACS naming conventions for member variables (trailing underscore).
```suggestion
  bool stop_called_{false};
```

eirrgang replied:
We also need to update the source file where this value is checked and used, changing it from a local variable to the member variable.

carolinedav replied:
Changed `stop_called` to `stop_called_` in second commit!

eirrgang commented on src/cpp/linearstoppotential.h:
```diff
@@ -58,6 +58,7 @@ class LinearStop {
                 const Resources &resources);
 
   double getTime() { return time_; }
+  bool getStop() { return stop_called_; }
```
You refer to `getStopCalled()` in the Python bindings. I think that's the better name.
```suggestion
  bool getStopCalled() { return stop_called_; }
```

carolinedav replied:
Changed `getStop` to `getStopCalled` in second commit!

eirrgang commented:
Well... it looks like Travis-CI finally stopped working for this repo. I'll update the automated testing in a separate commit. We can force the merge for this PR, if necessary, if it is ready before the automated testing is. Of course, we can and should test manually, too. :-)

eirrgang commented:
> wondering if we needed to include `stop_called` in the greater [class](https://github.com/kassonlab/brer_plugin/blob/baa32f0909314c21461ad30f973c82c0f3118e26/src/cpp/linearstoppotential.h#L44)

Actually, you need to expose the accessor (but not the data). Since LinearStopRestraint is the class that actually gets the Python bindings, and it privately inherits from LinearStop, add `using LinearStop::getStopCalled;` about 100 lines down, right after `using LinearStop::getTime;`.

carolinedav committed:
Updated linearstoppotential.h

Updated suggestions from pull request: changed stop_called to stop_called_ in initialization, added using LinearStop::getStopCalled, changed getStop() to getStopCalled()

carolinedav commented:
I added `using LinearStop::getStopCalled` in second commit

eirrgang reviewed (changes requested):
Note that we need to update linearstoppotential.cpp where it references `stop_not_called_`

carolinedav committed:
Updated `linearstoppotential.cpp` for a change in `stop_not_called` to `stop_called` and flipped logic to match

eirrgang commented on src/cpp/linearstoppotential.cpp:
```diff
@@ -78,8 +78,8 @@ void LinearStop::callback(gmx::Vector v, gmx::Vector v0, double t,
   if (converged) {
-    if (stop_not_called_) {
-      stop_not_called_ = false;
+    if !(stop_called_) {
```
```suggestion
    if (!stop_called_) {
```

eirrgang reviewed (changes requested):
This should have caused a compiler error if you tried to build it locally. It looks like the automated testing didn't catch it because you haven't merged `master` from `kassonlab/brer_plugin` since my updates yesterday.

Could you please fetch and merge `master` and push again _before_ accepting my suggested change? (I want to confirm that the automated testing is working right!)

eirrgang commented:
> Could you please fetch and merge `master` and push again _before_ accepting my suggested change? (I want to confirm that the automated testing is working right!)

Actually, there's probably a button in the GitHub interface to "bring your branch up to date" or something like that.

carolinedav committed:
Updated `linearstoppotential.cpp` to place 'not' inside if statement.
